src/Particular.ThroughputCollector.Persistence.RavenDb/ThroughputDataStore.cs

```diff
@@ -1,22 +1,131 @@
+
+        var documents = await query.ToListAsync(cancellationToken);
 
-        return endpoints.ToArray();
+        return documents.Select(document => document.ToEndpoint());
+    }
+
+    public async Task<Endpoint?> GetEndpoint(EndpointIdentifier id, CancellationToken cancellationToken = default)
+    {
+        using var session = store.OpenAsyncSession(databaseConfiguration.Name);
+
+        var documentId = id.GenerateDocumentId();
+
+        var document = await session.LoadAsync<EndpointDocument>(
+            documentId,
+            builder => builder.IncludeTimeSeries(ThroughputTimeSeriesName),
+            cancellationToken);
+
+        var endpoint = document?.ToEndpoint();
+        if (endpoint != null)
+        {
+            var timeSeries = await session
+                .IncrementalTimeSeriesFor(documentId, ThroughputTimeSeriesName)
+                .GetAsync(token: cancellationToken);
+
+            endpoint.LastCollectedDate = DateOnly.FromDateTime(timeSeries.LastOrDefault()?.Timestamp ?? DateTime.MinValue);
+        }
+
+        return endpoint;
+    }
+
+    public async Task<IEnumerable<(EndpointIdentifier, Endpoint?)>> GetEndpoints(IEnumerable<EndpointIdentifier> endpointIds, CancellationToken cancellationToken = default)
+    {
+        var endpoints = new List<(EndpointIdentifier, Endpoint?)>();
+
+        using var session = store.OpenAsyncSession(databaseConfiguration.Name);
+
+        var documentIdLookup = endpointIds.ToDictionary(
+            endpointId => endpointId,
+            endpointId => endpointId.GenerateDocumentId());
+
+        var endpointDocuments = await session.LoadAsync<EndpointDocument>(
+            documentIdLookup.Values.Distinct(),
+            builder => builder.IncludeTimeSeries(ThroughputTimeSeriesName),
+            cancellationToken);
+
+        foreach (var (documentId, endpointDocument) in endpointDocuments)
+        {
+            var id = endpointDocument == null
+             ? endpointIds.First(id => id.GenerateDocumentId().Equals(documentId))
+             : endpointDocument.EndpointId;
+
+            var endpoint = endpointDocument?.ToEndpoint();
+            if (endpoint != null)
+            {
+                var timeSeries = await session
+                    .IncrementalTimeSeriesFor(documentId, ThroughputTimeSeriesName)
+                    .GetAsync(token: cancellationToken);
+
+                endpoint.LastCollectedDate = DateOnly.FromDateTime(timeSeries.LastOrDefault()?.Timestamp ?? DateTime.MinValue);
+            }
+
+            endpoints.Add((id, endpoint));
+        }
+
+        return endpoints;
+    }
+
+    public async Task SaveEndpoint(Endpoint endpoint, CancellationToken cancellationToken = default)
+    {
+        var document = endpoint.ToEndpointDocument();
+
+        using var session = store.OpenAsyncSession("throughput");
+
+        await session.StoreAsync(document, document.GenerateDocumentId(), cancellationToken);
```

Comment:
So what happens here if the endpoint already exists in the db? I saw that in the InMemory persistence we do a check and remove it if it does.

Reply:
StoreAsync will overwrite it.